# Structured and time series data

This notebook contains an implementation of the third place result in the Rossman Kaggle competition as detailed in Guo/Berkhahn's [Entity Embeddings of Categorical Variables](https://arxiv.org/abs/1604.06737).

The motivation behind exploring this architecture is it's relevance to real-world application. Most data used for decision making day-to-day in industry is structured and/or time-series data. Here we explore the end-to-end process of using neural networks with practical structured data problems.

In [1]:
%matplotlib inline
%reload_ext autoreload
%autoreload 2

In [2]:
from fastai.structured import *
from fastai.column_data import *
np.set_printoptions(threshold=50, edgeitems=20)

PATH='data/rossmann/'

# Create datasets

In addition to the provided data, we will be using external datasets put together by participants in the Kaggle competition. You can download all of them [here](http://files.fast.ai/part2/lesson14/rossmann.tgz).

For completeness, the implementation used to put them together is included below.

In [3]:
def concat_csvs(dirname):
    path = f'{PATH}{dirname}'
    filenames=glob(f"{PATH}/*.csv")

    wrote_header = False
    with open(f"{path}.csv","w") as outputfile:
        for filename in filenames:
            name = filename.split(".")[0]
            with open(filename) as f:
                line = f.readline()
                if not wrote_header:
                    wrote_header = True
                    outputfile.write("file,"+line)
                for line in f:
                     outputfile.write(name + "," + line)
                outputfile.write("\n")

In [4]:
# concat_csvs('googletrend')
# concat_csvs('weather')

Feature Space:
* train: Training set provided by competition
* store: List of stores
* store_states: mapping of store to the German state they are in
* List of German state names
* googletrend: trend of certain google keywords over time, found by users to correlate well w/ given data
* weather: weather
* test: testing set

In [5]:
table_names = ['train', 'store', 'store_states', 'state_names', 
               'googletrend', 'weather', 'test']

We'll be using the popular data manipulation framework `pandas`. Among other things, pandas allows you to manipulate tables/data frames in python as one would in a database.

We're going to go ahead and load all of our csv's as dataframes into the list `tables`.

In [6]:
tables = [pd.read_csv(f'{PATH}{fname}.csv', low_memory=False) for fname in table_names]

In [7]:
from IPython.display import HTML

We can use `head()` to get a quick look at the contents of each table:
* train: Contains store information on a daily basis, tracks things like sales, customers, whether that day was a holdiay, etc.
* store: general info about the store including competition, etc.
* store_states: maps store to state it is in
* state_names: Maps state abbreviations to names
* googletrend: trend data for particular week/state
* weather: weather conditions for each state
* test: Same as training table, w/o sales and customers


In [ ]:
for n,t in zip(table_names,tables):
    display(t.head())
    print(f"Above is first 5 rows of {n.upper()} table with total {len(t):,} rows")

This is very representative of a typical industry dataset.

The following returns summarized aggregate information to each table accross each field.

In [ ]:
for n,t in zip(table_names,tables):
    display(DataFrameSummary(t).summary())
    print(f'{n.upper()} Table')

# Data Cleaning / Feature Engineering

As a structured data problem, we necessarily have to go through all the cleaning and feature engineering, even though we're using a neural network.

In [ ]:
# assign names to all table in tables list
# notice that looping not required
train, store, store_states, state_names, googletrend, weather, test = tables 

In [ ]:
len(train),len(test)

We turn state Holidays to booleans, to make them more convenient for modeling. We can do calculations on pandas fields using notation very similar (often identical) to numpy.

In [ ]:
# understand the diff values of StateHoliday, notice it's categorical not boolean or numeric
display(train.StateHoliday.unique())
display(test.StateHoliday.unique())

In [ ]:
# if not 0 (a,b,c), then it's True, else False
train.StateHoliday = train.StateHoliday!='0'
test.StateHoliday = test.StateHoliday!='0'

`join_df` is a function for joining tables on specific fields. By default, we'll be doing a left outer join of `right` on the `left` argument using the given fields for each table.

Pandas does joins using the `merge` method. The `suffixes` argument describes the naming convention for duplicate fields. We've elected to leave the duplicate field names on the left untouched, and append a "\_y" to those on the right.

In [ ]:
def join_df(left, right, left_on, right_on=None, suffix='_y'):
    if right_on is None: right_on = left_on
    return left.merge(right, how='left', left_on=left_on, right_on=right_on, 
                      suffixes=("", suffix))

Join weather/state names.

In [ ]:
# weather's file column corresponds to state_names' StateName column
weather = join_df(weather, state_names, "file", "StateName")

In [ ]:
weather.head() # notice both 'file' and 'StateName' column exist together

In pandas you can add new columns to a dataframe by simply defining it. We'll do this for googletrends by extracting dates and state names from the given data and adding those columns.

We're also going to replace all instances of state name 'NI' to match the usage in the rest of the data: 'HB,NI'. This is a good opportunity to highlight pandas indexing. We can use `.loc[rows, cols]` to select a list of rows and a list of columns from the dataframe. In this case, we're selecting rows w/ statename 'NI' by using a boolean list `googletrend.State=='NI'` and selecting "State".

In [ ]:
# before
display(googletrend.head(1))
display(pd.DataFrame(googletrend.iloc[1271]).T)

In [ ]:
googletrend['Date'] = googletrend.week.str.split(' - ', expand=True)[0] # extract date from beginning of week

In [ ]:
display(googletrend.head(1))
display(pd.DataFrame(googletrend.iloc[1271]).T)

In [ ]:
googletrend['State'] = googletrend.file.str.split('_', expand=True)[2] # extract State from file

In [ ]:
display(googletrend.head(1))
display(pd.DataFrame(googletrend.iloc[1271]).T)

In [ ]:
googletrend.loc[googletrend.State=='NI', "State"] = 'HB,NI'  
# selecting rows w/ statename 'NI' by using a boolean list googletrend.State=='NI' and selecting "State".

In [ ]:
pd.DataFrame(googletrend.iloc[1271]).T

The following extracts particular date fields from a complete datetime for the purpose of constructing categoricals.

You should *always* consider this feature extraction step when working with date-time. Without expanding your date-time into these additional fields, you can't capture any trend/cyclical behavior as a function of time at any of these granularities. We'll add to every table with a date field.

In [ ]:
add_datepart(weather, "Date", drop=False)
add_datepart(googletrend, "Date", drop=False)
add_datepart(train, "Date", drop=False)
add_datepart(test, "Date", drop=False)

The Google trends data has a special category for the whole of the Germany - we'll pull that out so we can use it explicitly.

In [ ]:
googletrend.file.unique()

In [ ]:
trend_de = googletrend[googletrend.file == 'Rossmann_DE']

Now we can outer join all of our data into a single dataframe. Recall that in outer joins everytime a value in the joining field on the left table does not have a corresponding value on the right table, the corresponding row in the new table has Null values for all right table fields. One way to check that all records are consistent and complete is to check for Null values post-join, as we do here.

*Aside*: Why note just do an inner join?
If you are assuming that all records are complete and match on the field you desire, an inner join will do the same thing as an outer join. However, in the event you are wrong or a mistake is made, an outer join followed by a null-check will catch it. (Comparing before/after # of rows for inner join is equivalent, but requires keeping track of before/after row #'s. Outer join is easier.)

In [ ]:
store = join_df(store, store_states, "Store")
len(store[store.State.isnull()])

In [ ]:
joined = join_df(train, store, "Store")
joined_test = join_df(test, store, "Store")
len(joined[joined.StoreType.isnull()]),len(joined_test[joined_test.StoreType.isnull()])

In [ ]:
joined = join_df(joined, googletrend, ["State","Year", "Week"])
joined_test = join_df(joined_test, googletrend, ["State","Year", "Week"])
len(joined[joined.trend.isnull()]),len(joined_test[joined_test.trend.isnull()])

In [ ]:
joined = joined.merge(trend_de, 'left', ["Year", "Week"], suffixes=('', '_DE'))
joined_test = joined_test.merge(trend_de, 'left', ["Year", "Week"], suffixes=('', '_DE'))
len(joined[joined.trend_DE.isnull()]),len(joined_test[joined_test.trend_DE.isnull()])

In [ ]:
joined = join_df(joined, weather, ["State","Date"])
joined_test = join_df(joined_test, weather, ["State","Date"])
len(joined[joined.Mean_TemperatureC.isnull()]),len(joined_test[joined_test.Mean_TemperatureC.isnull()])

In [ ]:
display(joined.head(1))
display(joined_test.head(1))

In [ ]:
# remove all the _y columns
for df in (joined, joined_test):
    for c in df.columns:
        if c.endswith('_y'):
            if c in df.columns: df.drop(c, inplace=True, axis=1)

When we call `for df in (joined, joined_test)`, we are actually repeating the procedure twice, first time for training data (joined) second time for test data (joined_test)

In [ ]:
display(joined.head(1))
display(joined_test.head(1))

Next we'll fill in missing values to avoid complications with `NA`'s. `NA` (not available) is how Pandas indicates missing values; many models have problems when missing values are present, so it's always important to think about how to deal with them. In these cases, we are picking an arbitrary *signal value* that doesn't otherwise appear in the data.

In [ ]:
for df in (joined,joined_test):
    df['CompetitionOpenSinceYear'] = df.CompetitionOpenSinceYear.fillna(1900).astype(np.int32)
    df['CompetitionOpenSinceMonth'] = df.CompetitionOpenSinceMonth.fillna(1).astype(np.int32)
    df['Promo2SinceYear'] = df.Promo2SinceYear.fillna(1900).astype(np.int32)
    df['Promo2SinceWeek'] = df.Promo2SinceWeek.fillna(1).astype(np.int32)

Next we'll extract features "CompetitionOpenSince" and "CompetitionDaysOpen". Note the use of `apply()` in mapping a function across dataframe values.

In [ ]:
for df in (joined,joined_test):
    df["CompetitionOpenSince"] = pd.to_datetime(dict(year=df.CompetitionOpenSinceYear, 
                                                     month=df.CompetitionOpenSinceMonth, day=15))
    df["CompetitionDaysOpen"] = df.Date.subtract(df.CompetitionOpenSince).dt.days

We'll replace some erroneous / outlying data.

In [ ]:
for df in (joined,joined_test):
    df.loc[df.CompetitionDaysOpen<0, "CompetitionDaysOpen"] = 0
    df.loc[df.CompetitionOpenSinceYear<1990, "CompetitionDaysOpen"] = 0

We add "CompetitionMonthsOpen" field, limiting the maximum to 2 years to limit number of unique categories.

In [ ]:
for df in (joined,joined_test):
    df["CompetitionMonthsOpen"] = df["CompetitionDaysOpen"]//30
    df.loc[df.CompetitionMonthsOpen>24, "CompetitionMonthsOpen"] = 24
joined.CompetitionMonthsOpen.unique()

Same process for Promo dates.

In [ ]:
#bottleneck 
for df in (joined,joined_test):
    df["Promo2Since"] = pd.to_datetime(df.apply(lambda x: Week(
        x.Promo2SinceYear, x.Promo2SinceWeek).monday(), axis=1).astype(pd.datetime))
    df["Promo2Days"] = df.Date.subtract(df["Promo2Since"]).dt.days

In [ ]:
for df in (joined,joined_test):
    df.loc[df.Promo2Days<0, "Promo2Days"] = 0
    df.loc[df.Promo2SinceYear<1990, "Promo2Days"] = 0
    df["Promo2Weeks"] = df["Promo2Days"]//7
    df.loc[df.Promo2Weeks<0, "Promo2Weeks"] = 0
    df.loc[df.Promo2Weeks>25, "Promo2Weeks"] = 25
    df.Promo2Weeks.unique()

In [ ]:
joined.to_feather(f'{PATH}joined_pre')
joined_test.to_feather(f'{PATH}joined_test_pre')

# Durations

In [ ]:
joined = pd.read_feather(f'{PATH}joined_pre')
joined_test = pd.read_feather(f'{PATH}joined_test_pre')

It is common when working with time series data to extract data that explains relationships across rows as opposed to columns, e.g.:
* Running averages
* Time until next event
* Time since last event

This is often difficult to do with most table manipulation frameworks, since they are designed to work with relationships across columns. As such, we've created a class to handle this type of data.

We'll define a function `get_elapsed` for cumulative counting across a sorted dataframe. Given a particular field `fld` to monitor, this function will start tracking time since the last occurrence of that field. When the field is seen again, the counter is set to zero.

Upon initialization, this will result in datetime na's until the field is encountered. This is reset every time a new store is seen. We'll see how to use this shortly.

In [ ]:
def get_elapsed(fld, pre):
    day1 = np.timedelta64(1, 'D')
    last_date = np.datetime64()
    last_store = 0
    res = []

    for s,v,d in zip(df.Store.values,df[fld].values, df.Date.values):
        if s != last_store:
            last_date = np.datetime64()
            last_store = s
        if v: last_date = d
        res.append(((d-last_date).astype('timedelta64[D]') / day1))
    df[pre+fld] = res

We'll be applying this to a subset of columns:

In [ ]:
columns = ["Date", "Store", "Promo", "StateHoliday", "SchoolHoliday"]

In [ ]:
columns

In [ ]:
# Run twice, first on train data, and then set run_on_train = False and run again

run_on_train = False

if run_on_train:
    df = train[columns]
else:
    df = test[columns]

Let's walk through an example.

Say we're looking at School Holiday. We'll first sort by Store, then Date, and then call `add_elapsed('SchoolHoliday', 'After')`:
This will apply to each row with School Holiday:
* A applied to every row of the dataframe in order of store and date
* Will add to the dataframe the days since seeing a School Holiday
* If we sort in the other direction, this will count the days until another holiday.

In [ ]:
fld = 'SchoolHoliday'
df = df.sort_values(['Store', 'Date'])
get_elapsed(fld, 'After')
df = df.sort_values(['Store', 'Date'], ascending=[True, False])
get_elapsed(fld, 'Before')

We'll do this for two more fields.

In [ ]:
fld = 'StateHoliday'
df = df.sort_values(['Store', 'Date'])
get_elapsed(fld, 'After')
df = df.sort_values(['Store', 'Date'], ascending=[True, False])
get_elapsed(fld, 'Before')

In [ ]:
fld = 'Promo'
df = df.sort_values(['Store', 'Date'])
get_elapsed(fld, 'After')
df = df.sort_values(['Store', 'Date'], ascending=[True, False])
get_elapsed(fld, 'Before')

We're going to set the active index to Date.

In [ ]:
df = df.set_index("Date")

Then set null values from elapsed field calculations to 0.

In [ ]:
columns = ['SchoolHoliday', 'StateHoliday', 'Promo']

In [ ]:
for o in ['Before', 'After']:
    for p in columns:
        a = o+p
        df[a] = df[a].fillna(0).astype(int)

Next we'll demonstrate window functions in pandas to calculate rolling quantities.

Here we're sorting by date (`sort_index()`) and counting the number of events of interest (`sum()`) defined in `columns` in the following week (`rolling()`), grouped by Store (`groupby()`). We do the same in the opposite direction.

In [ ]:
bwd = df[['Store']+columns].sort_index().groupby("Store").rolling(7, min_periods=1).sum()

In [ ]:
fwd = df[['Store']+columns].sort_index(ascending=False
                                      ).groupby("Store").rolling(7, min_periods=1).sum()

Next we want to drop the Store indices grouped together in the window function.

Often in pandas, there is an option to do this in place. This is time and memory efficient when working with large datasets.

In [ ]:
bwd.drop('Store',1,inplace=True)
bwd.reset_index(inplace=True)

In [ ]:
fwd.drop('Store',1,inplace=True)
fwd.reset_index(inplace=True)

In [ ]:
df.reset_index(inplace=True)

Now we'll merge these values onto the df.

In [ ]:
df = df.merge(bwd, 'left', ['Date', 'Store'], suffixes=['', '_bw'])
df = df.merge(fwd, 'left', ['Date', 'Store'], suffixes=['', '_fw'])

In [ ]:
df.drop(columns,1,inplace=True)

In [ ]:
df.head()

It's usually a good idea to back up large tables of extracted / wrangled features before you join them onto another one, that way you can go back to it easily if you need to make changes to it.

In [ ]:
#df.to_feather(f'{PATH}df')

In [ ]:
#df = pd.read_feather(f'{PATH}df')

In [ ]:
df["Date"] = pd.to_datetime(df.Date)

In [ ]:
df.columns

In [ ]:
if run_on_train:
    joined = join_df(joined, df, ['Store', 'Date'])
else:
    joined_test = join_df(joined_test, df, ['Store', 'Date'])

The authors also removed all instances where the store had zero sale / was closed. We speculate that this may have cost them a higher standing in the competition. One reason this may be the case is that a little exploratory data analysis reveals that there are often periods where stores are closed, typically for refurbishment. Before and after these periods, there are naturally spikes in sales that one might expect. By ommitting this data from their training, the authors gave up the ability to leverage information about these periods to predict this otherwise volatile behavior.

In [ ]:
# reset columns value and go back up and run again for test data by setting run_on_train=False
columns = ["Date", "Store", "Promo", "StateHoliday", "SchoolHoliday"]

In [ ]:
joined = joined[joined.Sales!=0]

We'll back this up as well.

In [ ]:
joined.reset_index(inplace=True)
joined_test.reset_index(inplace=True)

In [ ]:
joined.to_feather(f'{PATH}joined_no_zero_sales')
joined_test.to_feather(f'{PATH}joined_test_no_zero_sales')

We now have our final set of engineered features.

While these steps were explicitly outlined in the paper, these are all fairly typical feature engineering steps for dealing with time series data and are practical in any similar setting.

# Create features

In [8]:
# Begin from here next time once the final set of joined and joined_test are saved
# joined = pd.read_feather(f'{PATH}joined_all_sales')
# joined_test = pd.read_feather(f'{PATH}joined_test_all_sales')

joined = pd.read_feather(f'{PATH}joined_no_zero_sales')
joined_test = pd.read_feather(f'{PATH}joined_test_no_zero_sales')

## Handle the rows where sales are on the extreme ends - 0 and max

In [9]:
joined.Sales.quantile(q=0.0001),joined.Sales.quantile(0.01)

(762.3011, 2188.0)

In [10]:
joined.Sales.min(),joined.Sales.quantile(0.5), joined.Sales.max()

(46, 6369.0, 41551)

In [11]:
joined.Sales.quantile(0.99),joined.Sales.quantile(0.99999)

(17789.0, 37263.38402000163)

In [12]:
#joined.head().T.head(40)

## Continue

Now that we've engineered all our features, we need to convert to input compatible with a neural network.

This includes converting categorical variables into contiguous integers or one-hot encodings, normalizing continuous features to standard normal, etc...

In [13]:
len(joined.columns) # no. of features

93

In [14]:
cat_vars = ['Store', 'DayOfWeek', 'Year', 'Month', 'Day', 'StateHoliday', 'CompetitionMonthsOpen',
    'Promo2Weeks', 'StoreType', 'Assortment', 'PromoInterval', 'CompetitionOpenSinceYear', 'Promo2SinceYear',
    'State', 'Week', 'Events', 'Promo_fw', 'Promo_bw', 'StateHoliday_fw', 'StateHoliday_bw',
    'SchoolHoliday_fw', 'SchoolHoliday_bw']

contin_vars = ['CompetitionDistance', 'Max_TemperatureC', 'Mean_TemperatureC', 'Min_TemperatureC',
   'Max_Humidity', 'Mean_Humidity', 'Min_Humidity', 'Max_Wind_SpeedKm_h', 
   'Mean_Wind_SpeedKm_h', 'CloudCover', 'trend', 'trend_DE',
   'AfterStateHoliday', 'BeforeStateHoliday', 'Promo', 'SchoolHoliday']

n = len(joined); n

844338

844,338 instead of 1,017,209 because we've remove sales with 0 value

In [15]:
dep = 'Sales'
joined = joined[cat_vars+contin_vars+[dep, 'Date']].copy()

In [16]:
joined_test[dep] = 0 # broadcast zero for 'Sales' dummy col
joined_test = joined_test[cat_vars+contin_vars+[dep, 'Date', 'Id']].copy()

In [17]:
for v in cat_vars: joined[v] = joined[v].astype('category').cat.as_ordered()

In [18]:
apply_cats(joined_test, joined)

In [19]:
len(joined.columns), len(joined_test.columns)

(40, 41)

In [20]:
joined.columns

Index(['Store', 'DayOfWeek', 'Year', 'Month', 'Day', 'StateHoliday',
       'CompetitionMonthsOpen', 'Promo2Weeks', 'StoreType', 'Assortment',
       'PromoInterval', 'CompetitionOpenSinceYear', 'Promo2SinceYear', 'State',
       'Week', 'Events', 'Promo_fw', 'Promo_bw', 'StateHoliday_fw',
       'StateHoliday_bw', 'SchoolHoliday_fw', 'SchoolHoliday_bw',
       'CompetitionDistance', 'Max_TemperatureC', 'Mean_TemperatureC',
       'Min_TemperatureC', 'Max_Humidity', 'Mean_Humidity', 'Min_Humidity',
       'Max_Wind_SpeedKm_h', 'Mean_Wind_SpeedKm_h', 'CloudCover', 'trend',
       'trend_DE', 'AfterStateHoliday', 'BeforeStateHoliday', 'Promo',
       'SchoolHoliday', 'Sales', 'Date'],
      dtype='object')

In [21]:
joined_test.columns

Index(['Store', 'DayOfWeek', 'Year', 'Month', 'Day', 'StateHoliday',
       'CompetitionMonthsOpen', 'Promo2Weeks', 'StoreType', 'Assortment',
       'PromoInterval', 'CompetitionOpenSinceYear', 'Promo2SinceYear', 'State',
       'Week', 'Events', 'Promo_fw', 'Promo_bw', 'StateHoliday_fw',
       'StateHoliday_bw', 'SchoolHoliday_fw', 'SchoolHoliday_bw',
       'CompetitionDistance', 'Max_TemperatureC', 'Mean_TemperatureC',
       'Min_TemperatureC', 'Max_Humidity', 'Mean_Humidity', 'Min_Humidity',
       'Max_Wind_SpeedKm_h', 'Mean_Wind_SpeedKm_h', 'CloudCover', 'trend',
       'trend_DE', 'AfterStateHoliday', 'BeforeStateHoliday', 'Promo',
       'SchoolHoliday', 'Sales', 'Date', 'Id'],
      dtype='object')

In [22]:
for v in contin_vars:
    joined[v] = joined[v].fillna(0).astype('float32')
    joined_test[v] = joined_test[v].fillna(0).astype('float32')

We're going to run on a sample.

In [23]:
idxs = get_cv_idxs(n, val_pct=450000/n)
joined_samp = joined.iloc[idxs].set_index("Date")
samp_size = len(joined_samp); samp_size

450000

To run on the full dataset, use this instead:

In [24]:
samp_size = n
joined_samp = joined.set_index("Date")

We can now process our data...

In [25]:
joined_samp.head(2)

Store DayOfWeek  Year Month Day StateHoliday CompetitionMonthsOpen  \
Date                                                                            
2015-07-31     1         5  2015     7  31        False                    24   
2015-07-31     2         5  2015     7  31        False                    24   

           Promo2Weeks StoreType Assortment  ...  Max_Wind_SpeedKm_h  \
Date                                         ...                       
2015-07-31           0         c          a  ...                24.0   
2015-07-31          25         a          a  ...                14.0   

           Mean_Wind_SpeedKm_h CloudCover trend trend_DE AfterStateHoliday  \
Date                                                                         
2015-07-31                11.0        1.0  85.0     83.0              57.0   
2015-07-31                11.0        4.0  80.0     83.0              67.0   

           BeforeStateHoliday Promo SchoolHoliday Sales  
Date                                                     
2015-07-31                0.0   1.0           1.0  5263  
2015-07-31                0.0   1.0           1.0  6064  

[2 rows x 39 columns]

39 columns instead of 40 because Date has become index

In [26]:
len(joined_samp), len(joined_samp.columns)

(844338, 39)

In [27]:
df, y, nas, mapper = proc_df(joined_samp, 'Sales', do_scale=True)
yl = np.log(y)

In [28]:
joined_test = joined_test.set_index("Date")

In [29]:
df_test, _, nas, mapper = proc_df(joined_test, 'Sales', do_scale=True, skip_flds=['Id'],
                                  mapper=mapper, na_dict=nas)

In [30]:
len(joined_test), 

(41088,)

In [31]:
df.head(2)

Store  DayOfWeek  Year  Month  Day  StateHoliday  \
Date                                                           
2015-07-31      1          5     3      7   31             1   
2015-07-31      2          5     3      7   31             1   

            CompetitionMonthsOpen  Promo2Weeks  StoreType  Assortment  \
Date                                                                    
2015-07-31                     25            1          3           1   
2015-07-31                     25           26          1           1   

                ...        Min_Humidity  Max_Wind_SpeedKm_h  \
Date            ...                                           
2015-07-31      ...           -1.620066            0.149027   
2015-07-31      ...           -1.264031           -0.960613   

            Mean_Wind_SpeedKm_h  CloudCover     trend  trend_DE  \
Date                                                              
2015-07-31            -0.142774   -1.844823  1.732492  1.724334   
2015-07-31            -0.142774   -0.488722  1.294578  1.724334   

            AfterStateHoliday  BeforeStateHoliday     Promo  SchoolHoliday  
Date                                                                        
2015-07-31           0.604461            1.111827  1.113717        2.04105  
2015-07-31           0.926957            1.111827  1.113717        2.04105  

[2 rows x 38 columns]

38 columns down from 41 earlier - 'Sales' and 'id' dropped, 'Date' become index

In time series data, cross-validation is not random. Instead, our holdout data is generally the most recent data, as it would be in real application. This issue is discussed in detail in [this post](http://www.fast.ai/2017/11/13/validation-sets/) on our web site.

One approach is to take the last 25% of rows (sorted by date) as our validation set.

In [ ]:
# train_ratio = 0.75
# # train_ratio = 0.9
# train_size = int(samp_size * train_ratio); train_size
# val_idx = list(range(train_size, len(df)))

An even better option for picking a validation set is using the exact same length of time period as the test set uses - this is implemented here:

In [27]:
val_idx = np.flatnonzero(
    (df.index<=datetime.datetime(2014,9,17)) & (df.index>=datetime.datetime(2014,8,1)))

In [32]:
val_idx=[0]

In [33]:
len(val_idx)

1

# DL

We're ready to put together our models.

Root-mean-squared percent error is the metric Kaggle used for this competition.

In [34]:
def inv_y(a): return np.exp(a)

def exp_rmspe(y_pred, targ):
    targ = inv_y(targ)
    pct_var = (targ - inv_y(y_pred))/targ
    return math.sqrt((pct_var**2).mean())

max_log_y = np.max(yl)
y_range = (0, max_log_y*1.2)

In [35]:
def get_data():
    return ColumnarModelData.from_data_frame(PATH, val_idx, df, yl.astype(np.float32), cat_flds=cat_vars, bs=32,
                                       test_df=df_test)

We can create a ModelData object directly from out data frame.

In [36]:
md = get_data()

Some categorical variables have a lot more levels than others. Store, in particular, has over a thousand!

In [37]:
cat_sz = [(c, len(joined_samp[c].cat.categories)+1) for c in cat_vars]

In [38]:
cat_sz

[('Store', 1116),
 ('DayOfWeek', 8),
 ('Year', 4),
 ('Month', 13),
 ('Day', 32),
 ('StateHoliday', 3),
 ('CompetitionMonthsOpen', 26),
 ('Promo2Weeks', 27),
 ('StoreType', 5),
 ('Assortment', 4),
 ('PromoInterval', 4),
 ('CompetitionOpenSinceYear', 24),
 ('Promo2SinceYear', 9),
 ('State', 13),
 ('Week', 53),
 ('Events', 22),
 ('Promo_fw', 7),
 ('Promo_bw', 7),
 ('StateHoliday_fw', 4),
 ('StateHoliday_bw', 4),
 ('SchoolHoliday_fw', 9),
 ('SchoolHoliday_bw', 9)]

We use the *cardinality* of each variable (that is, its number of unique values) to decide how large to make its *embeddings*. Each level will be associated with a vector with length defined as below.

In [39]:
emb_szs = [(c, min(50, (c+1)//2)) for _,c in cat_sz]

In [40]:
emb_szs

[(1116, 50),
 (8, 4),
 (4, 2),
 (13, 7),
 (32, 16),
 (3, 2),
 (26, 13),
 (27, 14),
 (5, 3),
 (4, 2),
 (4, 2),
 (24, 12),
 (9, 5),
 (13, 7),
 (53, 27),
 (22, 11),
 (7, 4),
 (7, 4),
 (4, 2),
 (4, 2),
 (9, 5),
 (9, 5)]

In [41]:
m = md.get_learner(emb_szs, len(df.columns)-len(cat_vars),
                   0.04, 1, [1000,500], [0.001,0.01], y_range=y_range)

In [ ]:
m.lr_find()

In [ ]:
m.sched.plot(100)

In [42]:
lr = 1e-4

## Sample

In [ ]:
m.fit(lr, 3, metrics=[exp_rmspe])

In [ ]:
m.fit(lr, 5, metrics=[exp_rmspe], cycle_len=1)

In [ ]:
m.fit(lr, 2, metrics=[exp_rmspe], cycle_len=4)

In [ ]:
m.save('sample_trained')

In [ ]:
m.load('sample_trained')

In [ ]:
m.sched.plot_loss()

## All with validation

In [ ]:
md_all_2wk_val = get_data()

In [ ]:
m.set_data(md_all_2wk_val)

In [ ]:
m.fit(lr, 1, metrics=[exp_rmspe])

In [ ]:
m.fit(lr, 3, metrics=[exp_rmspe])

In [ ]:
m.fit(lr, 3, metrics=[exp_rmspe], cycle_len=1)

In [ ]:
m.sched.plot_loss()

In [ ]:
m.save('all_2wk_val_trained')

In [ ]:
m.load('all_2wk_val_trained')

## All without validation (validation merged into training)

In [38]:
m.fit(lr, 3, metrics=[exp_rmspe])

epoch      trn_loss   val_loss   exp_rmspe                                                                             
    0      0.012864   0.021372   0.128719  
    1      0.011148   0.013814   0.10806                                                                               
    2      0.009931   0.012793   0.102888                                                                              



[array([0.01279]), 0.10288765045328022]

In [39]:
m.fit(lr, 3, metrics=[exp_rmspe], cycle_len=1)

epoch      trn_loss   val_loss   exp_rmspe                                                                             
    0      0.007148   0.011541   0.098075  
    1      0.00668    0.011545   0.09803                                                                               
    2      0.008021   0.011492   0.098183                                                                              



[array([0.01149]), 0.09818287426305064]

In [ ]:
m.save('all_no_val_trained3')

In [ ]:
m.load('all_no_val_trained3')

## Test2 - Old (Private LB 0.11421)

In [43]:
md_all_no_val = get_data()

In [44]:
m.set_data(md_all_no_val)

In [45]:
# val_loss and metric does not matter here as we have set validation to only one sample
m.fit(lr, 3, metrics=[exp_rmspe])

epoch      trn_loss   val_loss   exp_rmspe                                                                             
    0      0.014223   0.007649   0.091396  
    1      0.011761   7e-06      0.00264                                                                               
    2      0.009539   7.1e-05    0.008371                                                                              



[array([0.00007]), 0.008371382048585889]

In [46]:
# val_loss and metric does not matter here as we have set validation to only one sample
m.fit(lr, 3, metrics=[exp_rmspe], cycle_len=1)

epoch      trn_loss   val_loss   exp_rmspe                                                                             
    0      0.007982   0.00032    0.017735  
    1      0.0084     0.000486   0.021804                                                                              
    2      0.007396   0.001287   0.035238                                                                                                                                  



[array([0.00129]), 0.0352376072658144]

In [47]:
m.save('sample_all_zero_val_trained')

In [48]:
m.load('sample_all_zero_val_trained')

## Resume here

In [49]:
x, y = m.predict_with_targs()

In [50]:
exp_rmspe(x,y)

0.0352376072658144

In [51]:
pred_test = m.predict(True)

In [52]:
pred_test = np.exp(pred_test)

In [53]:
joined_test['Sales'] = pred_test

In [54]:
csv_fn = f'{PATH}tmp/all_zero_val_trained_8.csv'

In [55]:
joined_test[['Id','Sales']].to_csv(csv_fn, index=False)

In [56]:
FileLink(csv_fn)

C:\Users\szelee\Projects\deep-learning\kaggle-bitbucket\rossmann-store-sales\data\rossmann\tmp\all_zero_val_trained_8.csv

1. Sample (subdata with 2 weeks validation)
1. Sample2 (Full data without validation data)
1. All (Almost full data with 2 weeks validation)
1. Test (Full data without validation data)

Kaggle Score: 

| Model | Private Score | Public Score |
| --- | --- | --- |
| Sample -> All -> Test (2 x 3epochs)  | 0.11497 | 0.10656  |
| Sample -> All -> Test (1 x 3epochs)  | 0.11784 |  0.10559 |
| Sample -> All  | 0.12633 | 0.10241 |
| Sample -> Test  | **0.11421<br>#108** | **0.10120<br>#145** |
| Sample -> Test  | 0.12423 | 0.10797 |
| Sample -> Test  | 0.11871 | 0.10647 |
| Sample2  | 0.12514 | 0.11969 |
| Test | 0.11615 | 0.10953 |

Jeremy's Howard Score 0.10754(Private #5) 0.10297(Public #340) [link](https://youtu.be/5_xFdhfUnvQ?t=2375)

# RF

In [ ]:
from sklearn.ensemble import RandomForestRegressor

In [ ]:
((val,trn), (y_val,y_trn)) = split_by_idx(val_idx, df.values, yl)

In [ ]:
m = RandomForestRegressor(n_estimators=40, max_features=0.99, min_samples_leaf=2,
                          n_jobs=-1, oob_score=True)
m.fit(trn, y_trn);

In [ ]:
preds = m.predict(val)
m.score(trn, y_trn), m.score(val, y_val), m.oob_score_, exp_rmspe(preds, y_val)